### PR TITLE
Update docs for BIP-5

### PR DESCRIPTION
**Pull request by dennwc**
* update BIP-5 description to more closely match final implementation (naming and syntax details)

* add a guide for transformation DSL

Work in progress.

### CODE REVIEW

File: driver/annotations.md [hunk 1]
```diff
@@ -1,17 +1,23 @@
 # Adding UAST Annotations
```
Comment (juanjux) [inline driver/annotations.md]: I think that the old annotation DSL should be removed from this file and probably be replaced (in this same file) by what you wrote on the transformation DSL file.

File: driver/transform-dsl.md [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+```
+
+For the "name" field the state is required to reconstruct it:
+
+```golang
+"name": Var("x"),
+```
+
+At this point, state variable "x" will be used to reconstruct the field
+value (assign it to null).
```
Comment (juanjux) [inline driver/transform-dsl.md]: maybe "and assign null to it" or "and set it to null".

File: driver/transform-dsl.md [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+a string, an array or another AST node.
+
+```json
+{
+    "type": "Ident",
+    "name": null
+}
+```
+
+Since there is no more constraints defined, `Part` transform will use
```
Comment (juanjux) [inline driver/transform-dsl.md]: **the** `Part` transform...

File: driver/transform-dsl.md [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+it's state stored in "other" variable to populate all unhandled fields:
```
Comment (juanjux) [inline driver/transform-dsl.md]: in **the ** "other" variable...

File: driver/transform-dsl.md [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+{
+    "type": "Ident",
+    "name": null,
+    "offset": 5
+}
+```
+
+As you can see, an operation can be used to describe a tree shape and
+reconstruct it back, assuming that all placeholder variables used in
+an operation are defined in node state.
```
Comment (juanjux) [inline driver/transform-dsl.md]: in **the** node state || in node's state.

File: driver/transform-dsl.md [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+s := NewState()
+
+// verify an AST with operation A, populate state
+if ok, _ := A.Check(s, ast); ok {
+    // construct a new AST using the state and operation B
+    newAST, _ := B.Construct(s, nil)
+}
+```
+
+As long as all variable defined in A exactly match all variables defined
```
Comment (juanjux) [inline driver/transform-dsl.md]: "as all variable**s**..."

File: driver/transform-dsl.md [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+in B, `Check`-`Construct` routine will work the same way as described
```
Comment (juanjux) [inline driver/transform-dsl.md]: **the** `Check` - `Construct` routine will work **in** the same way as described above (or in **the** example section above).

File: driver/transform-dsl.md [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+in example section above.
+
+Since state only stores placeholder values, while operation defines the
```
Comment (juanjux) [inline driver/transform-dsl.md]: Since **the** state... while **the** operation...

File: driver/transform-dsl.md [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+shape of the tree, any complex shape transformation can be defined:
+
+```golang
```
Comment (juanjux) [inline driver/transform-dsl.md]: Not review, just wanted to say that this example showing the code and the json nodes is brilliant and shows pretty clearly how the transformations and checks operate in a very simple way.

File: driver/transform-dsl.md [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+    },
+    "left": {"type":"Ident", "name": "v"},
+    "right": 5
+}
+```
+
+
+## Restrictions
+
+DSL also enforces few restrictions on transformations and operations.
```
Comment (juanjux) [inline driver/transform-dsl.md]: **The** DSL also enfoces **a** (or some) few restrictions...

File: driver/transform-dsl.md [hunk 11]
```diff
@@ -0,0 +1,549 @@
...
+### Completeness
+
+All operations should either:
+
+* Reject the tree if it does not match one of constraints.
+
+* Describe the whole tree, and save unknown/placeholder nodes to the state.
+
+This rule ensures that if an operation matches the tree, it would be able
+to replicate it from a state associated with the node, since all fields
```
Comment (juanjux) [inline driver/transform-dsl.md]: Missing comma after "that" and spurious comma after "node".

File: driver/transform-dsl.md [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+are covered by it.
+
+This property allows to test each transformation by executing
+`Check`-`Construct` routine, or detect if a mapping defined by a driver
```
Comment (juanjux) [inline driver/transform-dsl.md]: **the** `Check`-`Construct` routine... by a driver**'s** developer... (**which** would result in an error).

File: driver/transform-dsl.md [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+developer ignores certain AST fields that are seen in the wild
+(will result in an error).
+
+### Reversibility
+
+Since all operations are required to implement both `Check` and `Construct`
+they can be used interchangeably as either source and/or target transforms.
+
+In simple case it means that transform A can produce the state S for AST
```
Comment (juanjux) [inline driver/transform-dsl.md]: I would rephrase as:

"For a simple case, this means, that given a node T over which we apply a transform A, we can produce a state S (with `Check`) and then reconstruct the original node T (with `Construct`) by only using the state S."

File: driver/transform-dsl.md [hunk 14]
```diff
@@ -0,0 +1,549 @@
...
+node T (with `Check`), and then reconstruct an original node T (with `Construct`)
+by only using the state S.
+
+```
+S := A.Check(T); A.Construct(S) == T
+```
+
+In a more complete example, it's possible for transform A to check
```
Comment (juanjux) [inline driver/transform-dsl.md]: for **the** transform A (specially important in this case since without the the transform could be read as a verb).

File: driver/transform-dsl.md [hunk 15]
```diff
@@ -0,0 +1,549 @@
...
+the tree T, produce the state S and for transform B to use this state
```
Comment (juanjux) [inline driver/transform-dsl.md]: **which** produces...

File: driver/transform-dsl.md [hunk 16]
```diff
@@ -0,0 +1,549 @@
...
+T == A.Construct(S')
+```
+
+## List of operations
+
+### Values
+
+#### Any value
+
+In case any value can match the transformation, it should be stored into
```
Comment (juanjux) [inline driver/transform-dsl.md]: "If any value can...".

File: driver/transform-dsl.md [hunk 17]
```diff
@@ -0,0 +1,549 @@
...
+a named variable: `Var("x")`. Any name can be used for the variable as
+long as it is unique in this transformation.
+
+When writing mappings, variable with the same name should exist in the
```
Comment (juanjux) [inline driver/transform-dsl.md]: **a** variable with the same name

File: driver/transform-dsl.md [hunk 18]
```diff
@@ -0,0 +1,549 @@
...
+second transformation.
+
+#### Equal
+
+To check value for exact match, `Is(v)` operation is used.
+
+SDK defines helpers for common types: `String(v)`, `Int(v)`, etc.
```
Comment (juanjux) [inline driver/transform-dsl.md]: **The** SDK

File: driver/transform-dsl.md [hunk 19]
```diff
@@ -0,0 +1,549 @@
...
+
+#### Equal or nil
+
+To check any node for nil or specific rule `Opt` is used:
+
+```
+Opt("is_nil", Is(42))
+```
+
+`Opt` needs a named variable to store it's state ("is_nil" in this case).
```
Comment (juanjux) [inline driver/transform-dsl.md]: Mmmm this could use a small example. Also, Opt can be used not only for nil values but also for values that can be missing, right? If that's the case that should be written too.

File: driver/transform-dsl.md [hunk 20]
```diff
@@ -0,0 +1,549 @@
...
+
+#### Value in
+
+To check that value equal any value in the list, two operations should be
```
Comment (juanjux) [inline driver/transform-dsl.md]: that **a** value equal**s** (unsure if that's a verb, if not "is equal to").

File: driver/transform-dsl.md [hunk 21]
```diff
@@ -0,0 +1,549 @@
...
+executed:
+
+1) Check that value is in the list.
```
Comment (juanjux) [inline driver/transform-dsl.md]: that **the** value...

File: driver/transform-dsl.md [hunk 22]
```diff
@@ -0,0 +1,549 @@
...
+
+2) Store the value into a variable.
+
+To express this, `Check` helper can be used:
```
Comment (juanjux) [inline driver/transform-dsl.md]: **the** `Check` helper...

File: driver/transform-dsl.md [hunk 23]
```diff
@@ -0,0 +1,549 @@
...
+
+```golang
+Check(
+    // condition
+    In(
```
Comment (juanjux) [inline driver/transform-dsl.md]: Are the conditions inside the `In` evaluated as an "and" or "or"? In any case, that could be specified (my guess is that it's an "or" since the func is called "In" but being a little explicit won't hurt in this case).

File: driver/transform-dsl.md [hunk 24]
```diff
@@ -0,0 +1,549 @@
...
+Check(
+    // condition
+    Not(String("A")),
+
+    // store the value
+    Var("x"),
+)
+```
+
+Note that value can still be of any type and type constraint of `String`
```
Comment (juanjux) [inline driver/transform-dsl.md]: that **the** value... also the part after and "and" is  a little messy, maybe:

 "Note that the value of any type since we're using `Var` and thus the type constraint of `String` inside the `Check` will not affect the stored variable".

File: driver/transform-dsl.md [hunk 25]
```diff
@@ -0,0 +1,549 @@
...
+will not affect the variable.
+
+The same way a value can be checked for "not in" constraint:
```
Comment (juanjux) [inline driver/transform-dsl.md]: **In** the same way**,**... for **a/the**...

File: driver/transform-dsl.md [hunk 26]
```diff
@@ -0,0 +1,549 @@
...
+
+```
+Opt("is_nil", Obj{ ... })
+```
+
+`Opt` needs a named variable to store it's state ("is_nil" in this case).
+
+#### All fields must exist
+
+For object that always have a specific set of fields, `Obj` operation
```
Comment (juanjux) [inline driver/transform-dsl.md]: **the** `Obj`...

File: driver/transform-dsl.md [hunk 27]
```diff
@@ -0,0 +1,549 @@
...
+can be used:
+
+```golang
+Obj{
+    "type": String("Ident"),
+    "name": Var("name"),
+}
+```
+
+Since this operation is represented by a map, order of operations is not
```
Comment (juanjux) [inline driver/transform-dsl.md]: **the** order...

File: driver/transform-dsl.md [hunk 28]
```diff
@@ -0,0 +1,549 @@
...
+preserved. Operations will be executed in order of field name sorting.
```
Comment (juanjux) [inline driver/transform-dsl.md]: "will be executed following an alphabetical order from the field name"

File: driver/transform-dsl.md [hunk 29]
```diff
@@ -0,0 +1,549 @@
...
+This might affect operations that requres a variable to be defined, such
```
Comment (juanjux) [inline driver/transform-dsl.md]: requres => requires

File: driver/transform-dsl.md [hunk 30]
```diff
@@ -0,0 +1,549 @@
...
+as `Lookup` and `If`. To control execution order, use `Fields`.
```
Comment (juanjux) [inline driver/transform-dsl.md]: **the** execution order

File: driver/transform-dsl.md [hunk 31]
```diff
@@ -0,0 +1,549 @@
...
+
+#### Specific field options
+
+To get more control of how object fields are checked, use `Fields`:
```
Comment (juanjux) [inline driver/transform-dsl.md]: maybe: "about how **the** ordering and other aspects of the fields' checks..."

File: driver/transform-dsl.md [hunk 32]
```diff
@@ -0,0 +1,549 @@
...
+
+```golang
+Fields{
+    {Name: "type", Op: String("If")},
+    {Name: "then": Op: Var("then")},
+    {Name: "else": Op: Var("then"), Optional: "else_exists"},
+}
+```
+
+For each optional field a variable should be provided to store the state.
```
Comment (juanjux) [inline driver/transform-dsl.md]: Comma after field.

File: driver/transform-dsl.md [hunk 33]
```diff
@@ -0,0 +1,549 @@
...
+Optional field means that they may not exist. For checking nil fields see
+`Opt` operation.
+
```
Comment (juanjux) [inline driver/transform-dsl.md]: Since Field is pretty important, could use one of those expanded examples with the json representations of the objects before and after the transformation.

File: driver/transform-dsl.md [hunk 34]
```diff
@@ -0,0 +1,549 @@
...
+#### Skip unknown fields
+
+To skip unknown fields in object `Part` is used:
```
Comment (juanjux) [inline driver/transform-dsl.md]: I would rephrase as: `Part` can be used to skip unknown fields in an object:

File: driver/transform-dsl.md [hunk 35]
```diff
@@ -0,0 +1,549 @@
...
+
+```golang
+Part("other", Obj{
+    "name": Var("x"),
+})
+```
+
+`Part` will store all unhandled fields in `other` variable. It usually
```
Comment (juanjux) [inline driver/transform-dsl.md]: in **the** `other` variable. It usually required to have **a** `Part` call on both sides of a mapping (also, is this "usually" or always?).

File: driver/transform-dsl.md [hunk 36]
```diff
@@ -0,0 +1,549 @@
...
+requires to have `Part` on both sides of a mapping.
+
+### Arrays
+
+#### Fixed number of elements
+
+To represet an array with a specific number of elements, `Arr` can be used:
```
Comment (juanjux) [inline driver/transform-dsl.md]: Sounds more natural with the "`Arr` can be used..." at the start. Also, typo on represent.

File: driver/transform-dsl.md [hunk 37]
```diff
@@ -0,0 +1,549 @@
...
+
+```golang
+Arr(
+    // element 0, fixed
```
Comment (juanjux) [inline driver/transform-dsl.md]: Maybe expand a little with "// element 0, fixed, must be an int with a value of 42".

File: driver/transform-dsl.md [hunk 38]
```diff
@@ -0,0 +1,549 @@
...
+
+```golang
+Each("elems",
+    Obj{
+        "name": Var("x"),
+    },
+)
+```
+
+Note that `Each` requires a named variable to store elements. This also
```
Comment (juanjux) [inline driver/transform-dsl.md]: **the** elements.

File: driver/transform-dsl.md [hunk 39]
```diff
@@ -0,0 +1,549 @@
...
+means that `Each` should always be on both side of a mapping.
```
Comment (juanjux) [inline driver/transform-dsl.md]: side => sides.

File: driver/transform-dsl.md [hunk 40]
```diff
@@ -0,0 +1,549 @@
...
+
+`Each` allows an array to be empty or nil.
+
+Operation that is passed to `Each` will receive a clean state, thus
```
Comment (juanjux) [inline driver/transform-dsl.md]: **The** Operation that...

File: driver/transform-dsl.md [hunk 41]
```diff
@@ -0,0 +1,549 @@
...
+it will not be able to access any variables from the parent. It also
+means that variable names inside this operation will not collide with
+the variables defined by parent transformation.
+
```
Comment (juanjux) [inline driver/transform-dsl.md]: by **the** parent...

File: driver/transform-dsl.md [hunk 42]
```diff
@@ -0,0 +1,549 @@
...
+)
+```
+
+`Append` will verify that node is an array and will filter out "suffix"
+nodes before passing it to primary operation (`Var("x")` in this case).
+
+When used to costruct the tree, it will append "suffix" nodes to
+the end of an array defined by sub-operation.
+
+As in Go, `Append` will not affect variable "x".
```
Comment (juanjux) [inline driver/transform-dsl.md]: Not about this specific line: missing some doc and examples for `LookupArrOpVal` and friends.

File: proposals/bip-005.md [hunk 43]
```diff
@@ -88,35 +87,29 @@ to reason about the transformation in terms of writing nodes, as if
 a function call is not describing the structure, but constructing it:
 get variable `off`, convert to string, store it as the key `OffsetKey`.
 
-At the end, you can see a `OtherKeys` helper that will save all other
-node keys that were not used in previous transformations.
+You can see a `Part` helper wraps the main transformation. It will save
```
Comment (juanjux) [inline proposals/bip-005.md]: You can see **that**...

File: proposals/bip-005.md [hunk 44]
```diff
@@ -128,24 +121,20 @@ Note that this time str-to-int conversion is omitted - it only happens
 on one side of the transformation, because variable should store a
```
Comment (juanjux) [inline proposals/bip-005.md]: because **the/a** variable...

File: proposals/bip-005.md [hunk 45]
```diff
@@ -128,24 +121,20 @@ Note that this time str-to-int conversion is omitted - it only happens
...
 "canonical" representation of that value, as defined by final tree shape.
 
-Both tree shapes are provided as an argument to `Transformer` that will
-bind transformations together:
+Both tree shapes are provided as an argument to `Map` that will
```
Comment (juanjux) [inline proposals/bip-005.md]: Maybe:  as an argument => as arguments

File: proposals/bip-005.md [hunk 46]
```diff
@@ -128,24 +121,20 @@ Note that this time str-to-int conversion is omitted - it only happens
...
+bind transformations together creating a mapping:
```
Comment (juanjux) [inline proposals/bip-005.md]: bind **the**...

File: proposals/bip-005.md [hunk 47]
```diff
@@ -247,24 +235,21 @@ conversion, etc.
 * `Check(sel []Sel, op Op)` - checks all sel arguments before executing
   op's check. Reversal only executes apply of op. Mostly used for `Not`.
 
-* `Object(op Op)` - verifies that current node is an object and checks
-  it with provided op. Reversal changes node type to object and applies
-  a provided operation to it. This operation will populate a list of
+* `Obj map[string]Op` - verifies that current node is an object and checks
```
Comment (juanjux) [inline proposals/bip-005.md]: that **the** current...

**Comment by dennwc** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 75 unresolved discussions.

---

*[driver/annotations.md, line 1 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mivF1Vy3MnDUmr-r1-1:-LANStJB8OLX4bNzKKZa:b-5tyx40) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/annotations.md#L1)):*
<details><summary><i>Previously, juanjux (Juanjo Alvarez Martinez) wrote…</i></summary><blockquote>

I think that the old annotation DSL should be removed from this file and probably be replaced (in this same file) by what you wrote on the transformation DSL file.
</blockquote></details>

Yes, the plan was to update that section with a guide for new helpers as well. Still WIP.

---

*[driver/transform-dsl.md, line 29 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-29:-LANQarK4jrE2IfbphV9:b-hlgf0y) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L29)):*
<details><summary><i>Previously, juanjux (Juanjo Alvarez Martinez) wrote…</i></summary><blockquote>

This could be removed and the comments added to the snippet above.
</blockquote></details>

It was actually intentional :) Wanted to focus reader's attention one detail at a time - it might be easier to understand. But can revert it if you think it's okay in one snippet.

---

*[driver/transform-dsl.md, line 106 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-106:-LANWkJdDnah7Eq8SZfK:b-70p2n3) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L106)):*
> probably should also be specified that String does a check but also stores that given string in the shared state.

It's not :) Because this state is "static" (baked into this operation), thus it doesn't need to be stored.

---

*[driver/transform-dsl.md, line 136 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-136:-LANRMR4Bqf6PKqdAflk:bro00ib) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L136)):*
<details><summary><i>Previously, juanjux (Juanjo Alvarez Martinez) wrote…</i></summary><blockquote>

Part, if used, would replace `Map`, true? In that case this could be specified (something like "as partial, using a `Part` call instead of `Map` which...").
</blockquote></details>

No, it still requires a `Map`. Should I write the full snippet with `Map` to make it more clear?

---

*[driver/transform-dsl.md, line 374 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-374:-LANahCGE62_Yjq5GI6h:b-kb1p64) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L374)):*
> Also, Opt can be used not only for nil values but also for values that can be missing, right?

If by "missing values" you mean "missing fields", then no. There was a note above:

> Note that the DSL makes a distinction between null fields and fields that does not exist.

Now I added few more details. Is it more clear this way?

---

*[driver/transform-dsl.md, line 390 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-390:-LANc4oM1StjK62H8jou:bkas5sv) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L390)):*
<details><summary><i>Previously, juanjux (Juanjo Alvarez Martinez) wrote…</i></summary><blockquote>

Are the conditions inside the `In` evaluated as an "and" or "or"? In any case, that could be specified (my guess is that it's an "or" since the func is called "In" but being a little explicit won't hurt in this case).
</blockquote></details>

`In` is a specific helper that uses "or" logic, as a user would expect from a "value in ..." operation.
The `Check` itself has only one condition, so if it would be an `And`/`Or`, it would be mentioned explicitly.

---

*[driver/transform-dsl.md, line 489 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-489:-LANeCbLDXaebM0UHcJR:be4ia00) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L489)):*
<details><summary><i>Previously, juanjux (Juanjo Alvarez Martinez) wrote…</i></summary><blockquote>

in **the** `other` variable. It usually required to have **a** `Part` call on both sides of a mapping (also, is this "usually" or always?).
</blockquote></details>

Usually. Because you can have something like this:

```golang
Part("other", Obj{
  "name": Var("name"),
})
// ->
Obj{
  "name": Var("name"),
  "unrecognized": Var("other"),
}
```

Although I don't want to mention it because it's a bit more like implementation detail right now. If we will have a driver that relies on it, I will definitely describe it (and make it an API contract for `Part`).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/bblfsh/documentation/156)*

**Comment by juanjux** [comment]
*[driver/transform-dsl.md, line 29 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-29:-LAS7s3ZCqMEtwjUNFuc:b-dod832) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L29)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

It was actually intentional :) Wanted to focus reader's attention one detail at a time - it might be easier to understand. But can revert it if you think it's okay in one snippet.
</blockquote></details>

I think in this case is easier to grok the code with a well commented snippet.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/bblfsh/documentation/156)*

**Comment by juanjux** [comment]
Review status: 3 of 5 files reviewed at latest revision, 75 unresolved discussions.

---

*[driver/transform-dsl.md, line 136 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-136:-LAS9Bax9bRMnTGSR9Y0:bhzgyst) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L136)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

No, it still requires a `Map`. Should I write the full snippet with `Map` to make it more clear?
</blockquote></details>

Yes, I think it would be clearer that you need to do it inside a Map in that way.

---

*[driver/transform-dsl.md, line 374 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-374:-LAS9UBEBu7Lm9-9ZsXR:b8lgpgw) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L374)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

> Also, Opt can be used not only for nil values but also for values that can be missing, right?

If by "missing values" you mean "missing fields", then no. There was a note above:

> Note that the DSL makes a distinction between null fields and fields that does not exist.

Now I added few more details. Is it more clear this way?
</blockquote></details>

Yes, I think it's fine now.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/bblfsh/documentation/156)*

**Comment by juanjux** [comment]
Random idea: (this could be in a separate PR) it would be great to make a "recipes" page with some snippets like the ones you showed me these days to solve specific issues (doing different annotations depending on the value of a string, creating a new subnode to contain a list, etc.)

**Comment by juanjux** [comment]
Reviewed 1 of 5 files at r1, 1 of 2 files at r2.
Review status: 4 of 5 files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/bblfsh/documentation/156)*

**Comment by dennwc** [comment]
Was actually planning to add it to this PR with some examples of role annotations.

---

Reviewed 2 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-:-LAYeF5cAYNDEMCWUMKE:bivaz4q)*

**Comment by juanjux** [comment]
So, outside the snippets and other improvements, this is in pretty good shape but I guess we should wait for the SDK to be merged before merging this.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-:-LAlUpiC-jMKiW1IJT0W:bu5am1p)*

File: driver/transform-dsl.md (hunk 42 above)
Comment (juanjux) [reply driver/transform-dsl.md]: What do you think about this?

**Comment by dennwc** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[driver/transform-dsl.md, line 549 at r1](https://beta.reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-549:-LAljeMkDW2cxB7rtCST:b-lrinbk) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L549)):*
<details><summary><i>Previously, juanjux (Juanjo Alvarez Martinez) wrote…</i></summary><blockquote>

What do you think about this?
</blockquote></details>

Will add it a bit later - it's indeed a very important topic.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/bblfsh/documentation/156)*

**Comment by juanjux** [comment]
Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/bblfsh/documentation/156)*

**Comment by dennwc** [comment]
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[driver/annotations.md, line 163 at r4](https://reviewable.io/reviews/bblfsh/documentation/156#-LCEVldvD0IjKXf4vU38:-LCcTRVx9Wd9Za0ay8jX:bqj541m) ([raw file](https://github.com/bblfsh/documentation/blob/5d7e853672caa277e12709ea841f0f6ea04e338b/driver/annotations.md#L163)):*
> I'm not sure if the last part is true for this helper anymore
No, this helper is unordered, but it's OK.

---

*[driver/transform-dsl.md, line 478 at r1](https://reviewable.io/reviews/bblfsh/documentation/156#-LANO1mjl6BQ0BrSn9rN-r1-478:-LCcf58LCxKPk8546kqs:b-896fix) ([raw file](https://github.com/bblfsh/documentation/blob/8ad32e023b8943832c6c505803b2c170fb936eae/driver/transform-dsl.md#L478)):*
<details><summary><i>Previously, juanjux (Juanjo Alvarez Martinez) wrote…</i></summary><blockquote>

Since Field is pretty important, could use one of those expanded examples with the json representations of the objects before and after the transformation.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/bblfsh/documentation/156)*

**Comment by dennwc** [comment]
@juanjux I would say we should merge it as-is and add other docs as a separate PRs.

**Comment by juanjux** [comment]
There is a link that is broken (check CI), fix that and I'll happily merge.
